File: RobotLocalization.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
# @Date    : 2017-10-21 11:40:37
# @Author  : jingray (lionel_jing@163.com)
# @Link    : http://www.jianshu.com/u/01fb0364467d
# @Version : $Id$

import os

# Entropy=Σ(−p×log(p))
# In general, entropy represents the amount of uncertainty in a system.
# Since the measurement update step decreases uncertainty, entropy will decrease.
# The movement step increases uncertainty, so entropy will increase after this step.
p=[0.2, 0.2, 0.2, 0.2, 0.2]
world=['green', 'red', 'red', 'green', 'green']
measurements = ['red', 'red']
motions = [1,1]
pHit = 0.6
pMiss = 0.2
pExact = 0.8
pOvershoot = 0.1
pUndershoot = 0.1

def sense(p, Z):
    q=[]
    for i in range(len(p)):
        hit = (Z == world[i])
        q.append(p[i] * (hit * pHit + (1-hit) * pMiss))
    s = sum(q)
    for i in range(len(q)):
        q[i] = q[i] / s
    return q

def move(p, U):
    q = []
    for i in range(len(p)):
        s = pExact * p[(i-U) % len(p)]   # -1%5 = 4
        s = s + pOvershoot * p[(i-U-1) % len(p)]
        s = s + pUndershoot * p[(i-U+1) % len(p)]
        q.append(s)
    return q

for k in range(len(measurements)):
    p = sense(p, measurements[k])
    p = move(p, motions[k])

print(p)

File: Cplusplus/RobotLocalization.cpp
#include <iostream>
#include <vector>

using namespace std;

vector <float> p (5,0.2);
static const char worldarray[] = {'g', 'r', 'r', 'g', 'g'};
vector <char> worldvector (worldarray, worldarray+sizeof(worldarray)/sizeof(worldarray[0]));
static const char measurementsarray[] = {'r','r'};
vector <char> measurementvector (measurementsarray,measurementsarray+sizeof(measurementsarray)/sizeof(measurementsarray[0]));
//vector <char> measurementvector = {'r','r'};  /*support in version C++11*/
vector <int> motions (2,1);
float pHit = 0.6;
float pMiss = 0.2;
float pExact = 0.8;
float pOvershoot = 0.1;
float pUndershoot = 0.1;

vector <float> sense(vector <float> p, char Z)
{
	vector <float> q(p.size());
	bool hit;
	float sum = 0;
	for (int i = 0; i < p.size(); ++i)
	{
		hit = (Z == worldvector[i]);
		q[i] =p[i] * (hit * pHit + (1-hit) * pMiss);
		sum += q[i];
	}

	for (int i = 0; i < q.size(); ++i)
	{
		q[i] = q[i]/sum;
	}

	return q;
}

vector <float> move (vector <float>p, int U)
{
	vector <float> q(p.size());
	float qi;
	for (int i = 0; i < p.size(); ++i)
	{
		qi = pExact * p[(i-U+p.size())%p.size()];
		qi += pOvershoot * p[(i-U-1+p.size())%p.size()];
		qi += pUndershoot * p[(i-U+1+p.size())%p.size()];
		q[i] = qi;
	}
	return q;
}


int main(int argc, char const *argv[])
{
	vector <float> presult = p;

	for (int i = 0; i < measurementvector.size(); ++i)
	{
		presult = sense(presult,measurementvector[i]);
		presult = move(presult,motions[i]);
		/* code */
	}
	for (int i = 0; i < presult.size(); ++i)
	{
		cout << presult[i] << endl;
	}

	return 0;
}